### PR TITLE
Support for event id header

### PR DESCRIPTION
**Pull request by alexjch**
This header allows to group records by event_id.

### CODE REVIEW

File: src/telemetry.c [hunk 1]
```diff
@@ -637,6 +640,71 @@ static int set_bios_version_header(struct telem_ref *t_ref)
...
+ * @param t_ref Telemetry Record reference obtained from tm_create_record.
+ *
+ * @return 0 if successful, or a negative errno-style value if not.
+ *
+ */
+static int set_event_id_header(struct telem_ref *t_ref)
+{
+        int rc = 0;
+        int status = 0;
+        char buff[33] = {0};
```
Comment (icarus-sparry) [inline src/telemetry.c]: Not sure why there is both status and rc.

Not sure there is value in setting buff to 33*NUL just before setting it.

File: src/telemetry.c [hunk 2]
```diff
@@ -1016,6 +1090,55 @@ int tm_set_payload(struct telem_ref *t_ref, char *payload)
...
+ * @return 0 on success, 1 on failure
+ *
+ */
+static int validate_event_id(char *id)
+{
+       int rc = 0;
+       const char alphab[] = EVENT_ID_ALPHAB;
+
+       // 32 - 32 = 0
+       if((rc = (int)(strlen(id) - EVENT_ID_LEN)) != 0) {
```
Comment (bwarden) [inline src/telemetry.c]: Check for id==NULL before strlen

File: src/telemetry.c [hunk 3]
```diff
@@ -1016,6 +1090,55 @@ int tm_set_payload(struct telem_ref *t_ref, char *payload)
...
+/**
+ * Checks for id to have length = 32 characters
+ * and hexadecimal characters only.
+ *
+ * @param id A pointer to string to be checked
+ *
+ * @return 0 on success, 1 on failure
+ *
+ */
+static int validate_event_id(char *id)
```
Comment (bwarden) [inline src/telemetry.c]: Should this function be in a common place so it can be reused from telem_record_gen.c?

File: src/telemetry.c [hunk 4]
```diff
@@ -1016,6 +1090,58 @@ int tm_set_payload(struct telem_ref *t_ref, char *payload)
...
+       const char alphab[] = EVENT_ID_ALPHAB;
+
+       // 32 - 32 = 0
+       if((rc = (int)(strlen(id) - EVENT_ID_LEN)) != 0) {
+               return 1;
+       }
+
+       // verify alphabet
+       // strspn checks for characters in alphab
+       if(id[strspn(id, alphab)] != 0) {
```
Comment (icarus-sparry) [inline src/telemetry.c]: Since you already calculated strlen(id), you could just check that this is equal to the return value of strspn.

File: src/telemetry.c [hunk 5]
```diff
@@ -1016,6 +1090,55 @@ int tm_set_payload(struct telem_ref *t_ref, char *payload)
...
+}
+
+
+int tm_set_event_id(struct telem_ref *t_ref, char *event_id)
+{
+        int status = 0;
+
+        if(!(status = validate_event_id(event_id))) {
+                 // free default id before overriding
+                 free(t_ref->record->headers[TM_EVENT_ID]);
```
Comment (bwarden) [inline src/telemetry.c]: Check for t_ref == NULL and t_ref->record == NULL and t_ref->record->headers == NULL before this deref.

File: src/probes/telem_record_gen.c [hunk 6]
```diff
@@ -187,6 +193,17 @@ int validate_opts(void)
                 return ret;
         }
 
+        /* Event id  */
+        if(opt_event_id) {
+                if (strlen(opt_event_id) != 32) {
```
Comment (bwarden) [inline src/probes/telem_record_gen.c]: Limit should be defined in a common header file

File: tests/check_daemon.c [hunk 7]
```diff
@@ -298,7 +298,8 @@ START_TEST(check_process_record_with_correct_size_and_data)
                         "system_name: clear-linux-os\n"
                         "board_name: Qemu|Intel\n"
                         "cpu_model: Intel(R) Core(TM) i7-5650U CPU @ 2.20GHz\n"
-                        "bios_version: Qemu\n";
+                        "bios_version: Qemu\n"
+                        "event_id: 489d80c2f2aba1ff3c7530d0\n";
```
Comment (bwarden) [inline tests/check_daemon.c]: This event_id is only 24 characters, is that intentional?

File: src/telemetry.c [hunk 8]
```diff
@@ -15,6 +15,9 @@
  */
 
 #define _GNU_SOURCE
+
+#define EVENT_ID_LEN 32
```
Comment (bwarden) [inline src/telemetry.c]: Best defined in a single, common header file used in both client and server components.

File: src/probes/telem_record_gen.c [hunk 9]
```diff
@@ -140,6 +143,8 @@ int parse_options(int argc, char **argv)
 
                                 payload_version = (uint32_t)tmp;
                                 break;
+                        case 'e':
+                                opt_event_id = strdup(optarg);
```
Comment (icarus-sparry) [inline src/probes/telem_record_gen.c]: Maybe add a ```break;``` for consistency?

File: src/probes/telem_record_gen.c [hunk 10]
```diff
@@ -187,6 +193,17 @@ int validate_opts(void)
                 return ret;
         }
 
+        /* Event id  */
+        if(opt_event_id) {
+                if (strlen(opt_event_id) != 32) {
+                        fprintf(stderr, "Error: event_id should have 32 characters\n");
```
Comment (icarus-sparry) [inline src/probes/telem_record_gen.c]: You could report how many it had and what they were to avoid the user having to do so manually.

You could mention the need for the characters to be ascii, in this increasingly utf-8 world.

File: src/telemetry.c [hunk 11]
```diff
@@ -637,6 +640,71 @@ static int set_bios_version_header(struct telem_ref *t_ref)
...
+        frandom = open("/dev/urandom", O_RDONLY);
+        if (frandom < 0) {
+                goto rand_err;
+        }
+
+        if (read(frandom, &random_id, sizeof(random_id)) != sizeof(random_id)) {
+                goto rand_err;
+        }
+
+        result = sprintf(buff, "%.16" PRIx64 "%.16" PRIx64, random_id[0], random_id[1]);
```
Comment (bwarden) [inline src/telemetry.c]: I don't like getting a pointer here without some indication/protection of the buffer's length.

File: src/telemetry.c [hunk 12]
```diff
@@ -637,6 +640,71 @@ static int set_bios_version_header(struct telem_ref *t_ref)
...
+ * @return 0 if successful, or a negative errno-style value if not.
+ *
+ */
+static int set_event_id_header(struct telem_ref *t_ref)
+{
+        int rc = 0;
+        int status = 0;
+        char buff[33] = {0};
+
+        rc = gen_event_id(buff);
```
Comment (bwarden) [inline src/telemetry.c]: I think you should at least pass the size of buff here.

File: src/telemetry.c (hunk 1 above)
Comment (bwarden) [inline src/telemetry.c]: Size could come from common header.

File: src/telemetry.c [hunk 13]
```diff
@@ -637,6 +640,71 @@ static int set_bios_version_header(struct telem_ref *t_ref)
...
+
+        if (rc < 0) {
+                 status = rc;
+        } else {
+                 status = set_header(
+                        &(t_ref->record->headers[TM_EVENT_ID]),
```
Comment (bwarden) [inline src/telemetry.c]: Check each level for NULL before deref

File: src/telemetry.c [hunk 14]
```diff
@@ -637,6 +640,71 @@ static int set_bios_version_header(struct telem_ref *t_ref)
...
+                        TM_EVENT_ID_STR, buff,
+                        &(t_ref->record->header_size));
```
Comment (bwarden) [inline src/telemetry.c]: Check each level for NULL before deref

File: src/probes/telem_record_gen.c [hunk 15]
```diff
@@ -187,6 +193,17 @@ int validate_opts(void)
...
+                        return ret;
+                }
+                if(opt_event_id[strspn(opt_event_id, alphab)] != 0) {
+                        fprintf(stderr, "Error: event_id contains characters out of range, valid range %s\n", alphab);
```
Comment (icarus-sparry) [inline src/probes/telem_record_gen.c]: I think I would say

    Error: eventid contains invalid character '%c' at position %d in '%s', valid characters are %s

and fill in the 4 parameters. Makes the user do less work.

File: src/telemetry.c [hunk 16]
```diff
@@ -1016,6 +1090,58 @@ int tm_set_payload(struct telem_ref *t_ref, char *payload)
...
+ * @return 0 on success, 1 on failure
+ *
+ */
+static int validate_event_id(char *id)
+{
+       int rc = 0;
+       const char alphab[] = EVENT_ID_ALPHAB;
+
+       // 32 - 32 = 0
+       if((rc = (int)(strlen(id) - EVENT_ID_LEN)) != 0) {
```
Comment (tmarcu) [inline src/telemetry.c]: Isn't this defined to be 33 already?

File: tests/check_daemon.c (hunk 7 above)
Comment (icarus-sparry) [inline tests/check_daemon.c]: The test say "check_process_record_with_correct_size_and_data", but this has an invalidly sized event_id.

File: src/telemetry.h [hunk 17]
```diff
@@ -85,6 +85,19 @@ void tm_set_config_file(char *c_file);
...
+/**
+ * Sets the event_id to a user defined id
+ *
+ * @param t_ref A pointer to a telem_ref struct that was previously initialized
+ *        this struct will have an event_id already initialized and it will be
+ *        overriden by a user provided event_id
+ * @param event_id A pointer to a string that contains the 32 character length
+ *        event_id
+ *
+ * @return 0 on success, or s negative errno-style value on error
```
Comment (bwarden) [inline src/telemetry.h]: If I'm reading the functions right, this one always returns 0 or 1.

File: src/telemetry.c (hunk 4 above)
Comment (tmarcu) [inline src/telemetry.c]: PRIx64 prints only hexadecimal, so this validate doesn't need to exist unless you're getting an event ID from elsewhere.

File: tests/check_libtelemetry.c [hunk 18]
```diff
@@ -157,6 +157,45 @@ START_TEST(record_create_severity_overflow)
...
+                return;
+        }
+        ret = tm_set_event_id(ref, event_id);
+        ck_assert_int_eq(ret, 1);
+        ck_assert_str_ne(ref->record->headers[TM_EVENT_ID], result);
+        free(result);
+        create_teardown();
+}
+END_TEST
+
```
Comment (bwarden) [inline tests/check_libtelemetry.c]: Should also create tests for short event_id, long event_id, empty event_id, and NULL *event_id.

File: src/common.h [hunk 19]
```diff
@@ -53,14 +54,16 @@
 #define TM_BOARD_NAME_STR "board_name"
 #define TM_CPU_MODEL_STR "cpu_model"
 #define TM_BIOS_VERSION_STR "bios_version"
+#define TM_EVENT_ID_STR "event_id"
 
-#define NUM_HEADERS 14
+#define NUM_HEADERS 15
 
+#define EVENT_ID_ALPHAB "0123456789abcdef";
```
Comment (bwarden) [inline src/common.h]: Should we handle upper case hex chars too?

File: src/telemetry.c [hunk 20]
```diff
@@ -637,6 +640,71 @@ static int set_bios_version_header(struct telem_ref *t_ref)
         return status;
 }
 
+
+/**
+ * Generates an alphanumeric id of length len
```
Comment (tmarcu) [inline src/telemetry.c]: This generates an ID of length 32 exactly, update comment?

File: src/telemetry.c (hunk 4 above)
Comment (alexjch) [reply src/telemetry.c]: This id could be overridden from a public API tm_set_event_id, that's why I added this check.

File: src/probes/telem_record_gen.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (alexjch) [reply src/probes/telem_record_gen.c]: getopt check for the optargs parameter. But, I get your point I shiould not entirely rely on it. I'll add a check.

File: src/probes/telem_record_gen.c [hunk 21]
```diff
@@ -48,6 +49,7 @@ static const struct option prog_opts[] = {
         { "payload", required_argument, 0, 'p' },
         { "payload-file", required_argument, 0, 'P' },
         { "record-version", required_argument, 0, 'R' },
+        { "event-id", optional_argument, 0, 'e' },
```
Comment (phmccarty) [inline src/probes/telem_record_gen.c]: Shouldn't this be a required_argument? If the user wants a static event ID, I imagine they must supply a value.

File: src/probes/telem_record_gen.c [hunk 22]
```diff
@@ -67,6 +69,7 @@ static void print_help(void)
         printf("  -p, --payload         Record body (max size = 8k)\n");
         printf("  -P, --payload-file    File to read payload from\n");
         printf("  -R, --record-version  Version number for format of payload (default 1)\n");
+        printf("  -e  --event-id        Event id to use in the record\n");
```
Comment (phmccarty) [inline src/probes/telem_record_gen.c]: missing comma after -e

File: src/probes/telem_record_gen.c [hunk 23]
```diff
@@ -91,7 +94,7 @@ int parse_options(int argc, char **argv)
         long unsigned int tmp = 0;
 
         int opt;
-        while ((opt = getopt_long(argc, argv, "hc:Vs:c:p:P:R:", prog_opts, NULL)) != -1) {
+        while ((opt = getopt_long(argc, argv, "hc:Vs:c:p:P:R:e:", prog_opts, NULL)) != -1) {
```
Comment (phmccarty) [inline src/probes/telem_record_gen.c]: If the -e option is indeed optional, then you'll want to specify `e::` (two colons). If it is required, then `e:` (as-is) looks good.

File: src/probes/telem_record_gen.c [hunk 24]
```diff
@@ -187,6 +197,20 @@ int validate_opts(void)
                 return ret;
         }
 
+        /* Event id  */
+        if(opt_event_id) {
```
Comment (phmccarty) [inline src/probes/telem_record_gen.c]: style: space after `if`

File: src/telemetry.c [hunk 25]
```diff
@@ -637,6 +638,67 @@ static int set_bios_version_header(struct telem_ref *t_ref)
...
+        int result = -1;
+        uint64_t random_id[2] = {0};
+
+        frandom = open("/dev/urandom", O_RDONLY);
+        if (frandom < 0) {
+                return -1;
+        }
+
+        if (read(frandom, &random_id, sizeof(random_id)) == sizeof(random_id)) {
+                if(sprintf(buff, "%.16" PRIx64 "%.16" PRIx64, random_id[0], random_id[1]) == 32) {
```
Comment (phmccarty) [inline src/telemetry.c]: style: space after `if`

File: src/telemetry.c [hunk 26]
```diff
@@ -1016,6 +1084,60 @@ int tm_set_payload(struct telem_ref *t_ref, char *payload)
...
+ * @return 0 on success, 1 on failure
+ *
+ */
+static int validate_event_id(char *id)
+{
+       int rc = 0;
+       const char alphab[] = EVENT_ID_ALPHAB;
+
+       // make sure is not null
+       if(!id) {
```
Comment (phmccarty) [inline src/telemetry.c]: style: space after `if`

File: src/telemetry.c [hunk 27]
```diff
@@ -1016,6 +1084,60 @@ int tm_set_payload(struct telem_ref *t_ref, char *payload)
...
+               return 1;
+       }
+
+       // 32 - 32 = 0
+       if((rc = (int)(strlen(id) - EVENT_ID_LEN)) != 0) {
```
Comment (phmccarty) [inline src/telemetry.c]: style: space after `if`

File: src/telemetry.c [hunk 28]
```diff
@@ -1016,6 +1084,60 @@ int tm_set_payload(struct telem_ref *t_ref, char *payload)
...
+               return 1;
+       }
+
+       // verify alphabet
+       // strspn checks for characters in alphab
+       if(strspn(id, alphab) != EVENT_ID_LEN) {
```
Comment (phmccarty) [inline src/telemetry.c]: style: space after `if`

File: src/telemetry.c (hunk 5 above)
Comment (phmccarty) [reply src/telemetry.c]: I agree. It's better to be defensive here, since `tm_set_event_id` is the new API function.

File: src/telemetry.c [hunk 29]
```diff
@@ -1016,6 +1084,60 @@ int tm_set_payload(struct telem_ref *t_ref, char *payload)
...
+
+int tm_set_event_id(struct telem_ref *t_ref, char *event_id)
+{
+        int status = 0;
+
+        if(!(status = validate_event_id(event_id))) {
+                 // free default id before overriding
+                 free(t_ref->record->headers[TM_EVENT_ID]);
+                 // set new event_id
+                 if(asprintf(&(t_ref->record->headers[TM_EVENT_ID]),
```
Comment (phmccarty) [inline src/telemetry.c]: style: space after `if`

File: src/telemetry.h (hunk 17 above)
Comment (phmccarty) [reply src/telemetry.h]: I'm reading it this way too. @alexjch Should the function be returning a negative value, presuming that other public API functions do the same?

File: src/telemetry.sym [hunk 30]
```diff
@@ -15,4 +15,5 @@ TM_3_0_0 {
     tm_create_record;
     tm_set_payload;
     tm_send_record;
+    tm_set_event_id;
 } TM_2_0_0;
```
Comment (phmccarty) [inline src/telemetry.sym]: I think you should create a new `TM_4_0_0` block, and add `tm_set_event_id` to it with global visibility. This also means you should bump the library version to account for the new symbol. I think it's `SHAREDLIB_REVISION` you'll want to bump, but I would need to read the libtool docs for a refresher.

**Comment by phmccarty** [comment]
Also, it looks like Travis is failing...

File: src/probes/telem_record_gen.c (hunk 21 above)
Comment (alexjch) [reply src/probes/telem_record_gen.c]: Event id's are optional, if one is provided this will override a randomly generated during headers initialization.

**Comment by alexjch** [comment]
Travis was failing because source forge was in recovery mode due to DDOS

File: src/probes/telem_record_gen.c (hunk 21 above)
Comment (phmccarty) [reply src/probes/telem_record_gen.c]: Yes, but my point was that users won't pass -e/--event-id if they want a randomly generated ID. If they *do* pass the option, then I would expect an argument to be required.
